Comments: delete querystring action update

rodrigoi opened the pull request:
Part of #18321

When the comment view receives a `delete` command, it will redirect the user to the post view.

![delete](https://user-images.githubusercontent.com/233601/32808978-de3a3716-c973-11e7-89dd-1dbb8389d74a.gif)

How to test:

* Start Calypso with `ENABLE_FEATURES=comments/management/m3-design npm start`
* Navigate to a comment
* append `?action=delete` to the querystring.

gwwar commented on client/my-sites/comments/controller.js:
```diff
@@ -118,9 +118,11 @@ export const comment = ( { query, params, path, store } ) => {
 	}
 
 	const action = sanitizeQueryAction( query.action );
+	const redirectToPostView = postId =>
```
It looks like we pass through `siteFragment` so potentially we can create this function in Main and only pass it down one level.

rodrigoi replied:
I'd like to avoid giving the components direct access to `page`, and we'll have to pass it down the chain to declare this function on `main`.

gwwar commented:
Could you update testing instructions? Right now clicking on any action triggers the following error:
<img width="501" alt="screen shot 2017-11-14 at 4 21 10 pm" src="https://user-images.githubusercontent.com/1270189/32812073-f42cc94c-c957-11e7-8a34-2ea4676e1b95.png">

Copons commented:
@gwwar That's a somewhat expected issue.
`updateLastUndo()` comes from `CommentList`, and of course is undefined on single comment view.
But it's also not needed, so we should be good to go by just checking for its existence before calling it.

rodrigoi commented:
@gwwar @Copons sorry about the confusion. This PR affects only the delete action from the query string, not the _Trash_ or _Delete Permanently_ buttons. Those are coming later.

Copons commented:
@rodrigoi Yup, thanks for pointing it out!
I just wanted to observe that that error is a regression unrelated to this PR. 🙂

Copons commented:
It didn't occur to me until now: do we really _want_ to allow a (permanent and irreversible) `delete` action via URL?

When we perma-delete a comment via Calypso Comment Management, at least we get a confirm popup. In this case we'd just instantly delete the comment with no way back.

To be clear though: I don't think a confirm popup upon landing on Comment View should be the answer.
I actually think we should exclude `delete` from the URL actions.

Also, if I'm not mistaken, the moderation email only contains `approved`, `spam`, and `trash` as possible actions, so there's no actual _need_ for such action.

rodrigoi commented:
@copons that’s an excellent point. Maybe for delete we should add a warning with a confirm button. I’d like to hear what @megs and @drw158 think. 

The author notification email has a delete action. Maybe removing it altogether is an option.

gwwar commented:
Email moderation actions are "approve" "trash" or "mark as spam" we probably shouldn't allow a permanent delete, but trash in this case seems fine.

From this view folks _can_ permanently delete clicking on the comment action button, so we need to be able to redirect away when that happens.

rodrigoi commented:
ok, after further inspection, the option to delete comments via email link is not working on `wpcom` (code here 1c3a7-pb).
Since this is new for Jetpack, we are doing the same there: https://github.com/Automattic/jetpack/pull/8144/files#diff-3433271820eba273a5e94e1d9cd1a591R128

I think we should support this option since on Jetpack the user can redefine the `EMPTY_TRASH_DAYS` constant.

rodrigoi commented:
@megsfulton @drw158 I've added a small warning using a `Notice` (tried to use a [`Banner`](https://wpcalypso.wordpress.com/devdocs/design/banner), but it's too tied to plans).

The notice icon is a bit funky because of the redesign, we can remove it for now. To see the notice, navigate to a comment and add `?action=delete` to the URL.

![screen shot 2017-11-17 at 11 45 52](https://user-images.githubusercontent.com/233601/32952792-0e0df138-cb8d-11e7-8fe2-ce29f1d6c3a9.png)

@Copons @gwwar I've updated the `Delete Permanently` button to redirect to the _Post View_. I'm not very happy with the results, mostly because the timing of the redirect and the success/error notification, but I think we'll have to live with that until we have a way to handle side effects on the framework.
The only thing that it's a bit weird is that we always redirect to the _Post View_, but the user can get to the _Comment View_ from both the _Site_ and _Post_ views. I'll try fixing it with `history.back();`.

gwwar commented on client/my-sites/comments/comment/index.jsx:
```diff
@@ -30,6 +30,7 @@ export class Comment extends Component {
 		isPersistent: PropTypes.bool,
 		isPostView: PropTypes.bool,
 		isSelected: PropTypes.bool,
+		redirect: PropTypes.func,
```
Maybe `deleteRedirect`?

gwwar commented on client/my-sites/comments/controller.js:
```diff
@@ -118,9 +118,11 @@ export const comment = ( { query, params, path, store } ) => {
 
 	renderWithReduxStore(
-		<CommentView { ...{ action, commentId, siteFragment } } />,
+		<CommentView { ...{ action, commentId, siteFragment, redirectToPostView } } />,
```
I'd stick with one param name, but with a named value, eg:
```
deleteRedirect: redirectToPostView
```
So at a glance we know someone reading can guess that we redirect to post view, after a delete.

gwwar commented:
Overall I don't mind the notice up top. I'd recommend only displaying the "are you sure notice", if the comment has already been trashed.

I noticed some small text mis-alignment on the notice, but it might be fixed with a master rebase.
![screen shot 2017-11-20 at 2 46 19 pm](https://user-images.githubusercontent.com/1270189/33045493-d565852c-ce01-11e7-8ca1-825828a4f3f8.png)

@drw158 Maybe we need a Trash label here for the single comments view?

drw158 commented:
I like the idea of adding "are you sure" if they hit the Delete button in the email.

@rodrigoi perhaps the new cards with the left border could work https://wpcalypso.wordpress.com/devdocs/design/cards

I kind of like that the notice is more attention grabbing though.

> Maybe we need a Trash label here for the single comments view?
@gwwar What's the context on that? The user is on the single comment view and they hit Trash on an Approved comment?

Might be nice, but I think that would be making an exception for this view (all other times there is no confirmation on trash).

gwwar commented:
> @gwwar What's the context on that? The user is on the single comment view and they hit Trash on an Approved comment?

Try trashing something from approved. Go into the single comment view. How do I know that the comment is already trashed, but not permanently deleted? The actions change a bit, but it's not super obvious.

![screen shot 2017-11-20 at 3 31 04 pm](https://user-images.githubusercontent.com/1270189/33046904-e5250504-ce07-11e7-964c-7d5740f49a44.png)

drw158 commented:
Ah, right ok. I assume that we need this view, for trashing comments from an email? Just asking because I noticed that this view is not available for trashed comments in WP Admin. Maybe we could send them to the trash folder view? 

Here is what I suggest for Spam comments in this view, and could be applied to trash comments

https://github.com/Automattic/wp-calypso/issues/19773#issuecomment-345330332

drw158 commented:
Also sorry, still wrapping my head around this. Just wanted to clarify that I think

- we should not delete permanently via URL
- when you trash via email, there should be no confirm, it should just go to trash

Also, just noticed that when we delete permanently we don't show a confirm or undo button? Feels like we should have some sort of confirm maybe. Not sure what we decided on that before I left.

Copons commented:
> - we should not delete permanently via URL

+1

- Also, just noticed that when we delete permanently we don't show a confirm or undo button?

@drw158 Damn! This is a regression on the M4 design! I'll open an issue.

rodrigoi commented:
@drw158 supporting _Delete Permanently_ actions via email is a corner case for Jetpack sites. On Jetpack sites, the user has the option to [disable the _Trash_ feature for all content types](https://codex.wordpress.org/Trash_status), including comments:

| Before | After |
| -- | -- |
| ![screen shot 2017-11-21 at 13 46 55](https://user-images.githubusercontent.com/233601/33089535-a960053c-cecf-11e7-8bc2-2fcede722cf7.png) | ![screen shot 2017-11-21 at 13 47 22](https://user-images.githubusercontent.com/233601/33089548-b09a10c2-cecf-11e7-8612-4507d8a61c68.png) |

When _Trash_ is disabled, the comment moderation and the author notification emails include a link with the `delete` action, to permanently delete the comment:

| Author Notification | Comment Moderation |
| -- | -- |
| ![screen shot 2017-11-21 at 15 24 04](https://user-images.githubusercontent.com/233601/33089715-4291632c-ced0-11e7-8c1a-95e9a223f591.png)| ![screen shot 2017-11-21 at 15 24 32](https://user-images.githubusercontent.com/233601/33089724-4b052200-ced0-11e7-8c8b-7c14a14814ed.png) |

WP-Admin on WordPress.com has no _Permanently Delete_ option via email.

gwwar commented:
Unless folks found anything blocking, let's land this and open another PR/Issue with any follow up items.
